Implement s3-based spooling

linzebing opened the pull request:
This only includes the `trino-exchange` repo. (as branch `implement-task-level-retries` will constantly get updated until it gets merged, so only copy the trino-exchange repo for the ease of reviewers)

Previous code review see https://github.com/arhimondr/trino/pull/4

linzebing committed:
Implement s3-based spooling

losipiuk commented:
I commented https://github.com/arhimondr/trino/pull/4#discussion_r762798878 on the #4

losipiuk commented:
@linzebing Is there a way to see the changes you made in #4 while addressing comments. As you changed the base I have no idea how to do that. 
Can we try to keep base of PR constant (or rarely changing) through the life of the PR?

arhimondr commented on plugin/trino-exchange/pom.xml:
```diff
@@ -0,0 +1,165 @@
+        <dependency>
+            <groupId>io.trino</groupId>
+            <artifactId>trino-main</artifactId>
+            <type>test-jar</type>
```
Why is this jar needed?

linzebing replied:
It's not needed. It's a leftover when I copy-pasted the pom.xml content that I forgot to clean up.

arhimondr commented on plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSinkHandle.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(queryId, stageId, taskPartitionId);
```
`hashCode` doesn't include the `secretKey`, while it is still correct it could be a little confusing. I wonder if `equals` and `hashCode` methods are actually needed here? Have you considered storing the `taskPartitionId` in the `allSinks`/`finishedSinks` sets instead?

linzebing replied:
Makes sense

linzebing replied:
Makes sense. BTW I realized that queryId and stageId are not needed in `FileSystemExchangeSinkHandle`. Will just remove them.

arhimondr commented on plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSinkHandle.java:
```diff
@@ -0,0 +1,110 @@
+                .add("queryId", queryId)
+                .add("stageId", stageId)
+                .add("taskPartitionId", taskPartitionId)
+                .add("secretKey", secretKey) // TODO: understand the best practice here
```
maybe something like `secretKey.map(value -> "[redacted]")`?

arhimondr commented on plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSource.java:
```diff
@@ -0,0 +1,118 @@
+
+        if (sliceInput == null) {
+            if (files.hasNext()) {
+                URI file = files.next();
```
Reading one file at a time could be slow. File systems (such as S3) are usually backed by spinning disks with a limited throughput at 100 - 200 MB/s. This is likely to become a bottleneck. Ideally the implementation should open multiple files at a time (number of files can be configurable / adaptive) and read multiple files in parallel.

This doesn't have to be addressed in this PR and can be done as a follow up. Just something to keep in mind.

linzebing replied:
I will mark a TODO here. We can talk more tomorrow.

arhimondr commented on plugin/trino-exchange/pom.xml:
```diff
@@ -0,0 +1,165 @@
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>trino-exchange</artifactId>
```
Don't forget to add this module to the list of modules in the base pom

linzebing replied:
Yes it has been added (this PR only includes the trino-exchange plugin changes)

arhimondr commented on plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java:
```diff
@@ -0,0 +1,290 @@
+import static io.trino.plugin.exchange.FileSystemExchangeSink.DATA_FILE_SUFFIX;
+import static java.util.Objects.requireNonNull;
+
+public class FileSystemExchange
```
The `LocalFileSystemExchange` could be removed and the `FileSystemExchange` can be used for testing with a local file system.

linzebing replied:
Yes, those changes will be included in the future formal PR.

This PR only includes changes of trino-exchange plugin to avoid the hassle of staying in sync with branch `implement-task-level-retries`.

arhimondr commented on plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java:
```diff
@@ -0,0 +1,290 @@
+    @Override
+    public ExchangeSourceSplitter split(ExchangeSourceHandle handle, long targetSizeInBytes)
+    {
+        // TODO: replace with correct split implementation
```
Don't forget to update the implementation to respect the `targetSizeInBytes` parameter

linzebing replied:
QQ: do we still need this splitter? As we discussed, `StageTaskSourceFactory` will always split the files according to `targetSizeInBytes`. Does that mean we don't need this splitter anymore?

arhimondr replied:
IIUC the current implementation will produce one split per file. If the files are tiny it might not be necessary to produce as many partitions as there are files.

arhimondr commented on plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeStorage.java:
```diff
@@ -0,0 +1,49 @@
+
+    boolean exists(URI file);
+
+    void createFile(URI file) throws IOException;
```
This can be implemented as `getSliceInput(URI file, Optional<SecretKey> secretKey).close()`. Not sure if there's a value in adding a separate method to create an empty file.

linzebing replied:
Yes but that will create unnecessary memory allocation (and get GC'ed later). Since the implementation is basically one-liner, I would prefer to avoid memory allocation in this method.

arhimondr replied:
Sounds good. Maybe rename it to `createEmptyFile`?

arhimondr commented on plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeStorage.java:
```diff
@@ -0,0 +1,49 @@
+
+    Stream<URI> listDirectories(URI dir) throws IOException;
+
+    long size(URI file) throws IOException;
```
nit: `getSizeInBytes`

arhimondr commented on plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeStorage.java:
```diff
@@ -0,0 +1,49 @@
+
+    SliceOutput createSliceOutput(URI file, Optional<SecretKey> secretKey) throws IOException;
+
+    boolean exists(URI file);
```
Why every other method throws an `IOException`, but this one doesn't?

linzebing replied:
Because if the file path doesn't exist, we return false? Other methods will throw if path is not valid.

arhimondr replied:
In theory it may still throw (if there's a communication problem with a remote file system)

linzebing replied:
@arhimondr : S3 APIs throws SdkClientException (which extends RuntimeException) in such cases

arhimondr replied:
Do you think we should re-wrap S3 client specific exceptions into an `IOException` to avoid abstraction leaks?

linzebing replied:
Let me think about this. Currently I don't handle `SdkClientException` and `AmazonServiceException`  at all.

arhimondr commented on plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeStorage.java:
```diff
@@ -0,0 +1,49 @@
+
+    SliceInput getSliceInput(URI file, Optional<SecretKey> secretKey) throws IOException;
+
+    SliceOutput createSliceOutput(URI file, Optional<SecretKey> secretKey) throws IOException;
```
I'm not sure if at this level the implementation should be required to return a `SliceOutput`. That creates an additional implementation overhead. Have you considered adding a simple custom interface, something like

```
interface ExchangeStorageWriter {
  
  ListenableFuture<?> write(Slice slice);
  
  void close();

}
```

Such interface would also allow implementing an asynchronous IO operations (if the underlying file system supports an async interface). WDYT?

arhimondr replied:
nit: maybe simply `create`?

arhimondr commented on plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeStorage.java:
```diff
@@ -0,0 +1,49 @@
+{
+    void createDirectories(URI dir) throws IOException;
+
+    SliceInput getSliceInput(URI file, Optional<SecretKey> secretKey) throws IOException;
```
Same here. How about a simpler interface, e.g.:

```
interface ExchangeStorageReader {
  
  ListenableFuture<Integer> read(Slice destination, int offset, int size);
  
  void close();

}
```

?

arhimondr commented on plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/local/LocalFileSystemExchangeStorage.java:
```diff
@@ -0,0 +1,144 @@
+    public void createDirectories(URI dir)
+            throws IOException
+    {
+        Files.createDirectories(Paths.get(dir.getPath()));
```
nit: static import (here and in other places)

arhimondr commented on plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -0,0 +1,698 @@
+        private Future<UploadPartResult> inProgressUploadFuture;
+        private final List<UploadPartResult> parts = new ArrayList<>();
+
+        // TODO: implement async upload (aka. double buffering)
```
Not sure if this is the right level to implement async upload (double buffering). I wonder if it would be better to implement it at the `ExchangeSink` level. That way the mechanism could be reused for all future supported file systems at no additional cost.

arhimondr commented on plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java:
```diff
@@ -0,0 +1,698 @@
+                sseCustomerKey.ifPresent(request::withSSECustomerKey);
+
+                try {
+                    s3.putObject(request);
```
Does S3 client offer a non blocking (async) method?

linzebing replied:
Yes, however I need some time researching it on how much benefit it will bring to us.

linzebing commented:
> @linzebing Is there a way to see the changes you made in #4 while addressing comments. As you changed the base I have no idea how to do that. Can we try to keep base of PR constant (or rarely changing) through the life of the PR?

@losipiuk : sorry about the change, I have to request your re-review on this PR. Future changes will stick to this PR (as this PR is against master). #4 was against `implement-task-level-retries` which is being constantly updated, which brings the trouble to making my branch in sync.

linzebing commented:
Need to close it in favor of https://github.com/arhimondr/trino/pull/7 because the previous branch was gone as a result of my work on https://github.com/trinodb/trino/pull/10180. I promise I will not change base again.

losipiuk commented:
> Need to close it in favor of #7 because the previous branch was gone as a result of my work on [trinodb#10180](https://github.com/trinodb/trino/pull/10180). I promise I will not change base again.

:)

FWIW I think you can change the base without opening a new PR.

![image](https://user-images.githubusercontent.com/890111/145284175-1a799828-d26e-4426-ab6e-4e387a9b4a2b.png)
